Add additional tracker info to WebUI

Piccirello opened the pull request:
Note: This does not break anything in the existing `/torrents/trackers` API; it only adds additional info


Before:
<img width="785" alt="screen shot 2018-08-20 at 12 15 27 am" src="https://user-images.githubusercontent.com/8296030/44322007-3a0f0280-a419-11e8-85f6-60f4eb881fc3.png">

After:
<img width="826" alt="screen shot 2018-08-20 at 1 31 46 am" src="https://user-images.githubusercontent.com/8296030/44322013-3c715c80-a419-11e8-9880-4a775d2b91b9.png">

GUI, for comparison:
<img width="865" alt="screen shot 2018-08-20 at 12 33 03 am" src="https://user-images.githubusercontent.com/8296030/44322017-41361080-a419-11e8-85b3-ac3284bc3834.png">


**Context menu:**

webui: 
<img width="493" alt="screen shot 2018-09-02 at 6 35 26 pm" src="https://user-images.githubusercontent.com/8296030/44961425-1353d000-aedf-11e8-8168-46fd2c004a01.png">


gui: 
<img width="583" alt="screen shot 2018-09-02 at 6 44 04 pm" src="https://user-images.githubusercontent.com/8296030/44961520-477bc080-aee0-11e8-8797-48f0b755a5d2.png">


**Category editing**

webui:
<img width="514" alt="screen shot 2018-09-02 at 6 30 49 pm" src="https://user-images.githubusercontent.com/8296030/44961384-65e0bc80-aede-11e8-8ff2-43be744f4c48.png">

gui:
<img width="380" alt="screen shot 2018-09-02 at 6 31 52 pm" src="https://user-images.githubusercontent.com/8296030/44961393-8c065c80-aede-11e8-9c9b-6d316143ecc0.png">

- [x] Update WebAPI documentation

Closes #7934, closes #6797, closes #3024

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -304,6 +312,8 @@ void TorrentsController::trackersAction()
     if (!torrent)
         throw APIError(APIErrorType::NotFound);
 
+    trackerList << getStickyTrackers(torrent);
```
`QVariantList trackerList {getStickyTrackers(torrent)};`

Piccirello replied:
This change results in the `getStickyTrackers` QList being added as an element of `trackerList`, rather than the elements of `getStickyTrackers` being added.

Chocobo1 replied:
oops, it is because using the braces, can you give it one last try: 
```c++
QVariantList trackerList = getStickyTrackers(torrent);
```

Piccirello replied:
That worked!

Chocobo1 commented on src/webui/api/torrentscontroller.h:
```diff
@@ -30,6 +30,8 @@
 
 #include "apicontroller.h"
 
+#include "base/bittorrent/torrenthandle.h"
```
Forward declare it instead: `class BitTorrent::TorrentHandle;` and revert this and also in cpp file.

Chocobo1 commented on src/webui/api/torrentscontroller.h:
```diff
@@ -73,4 +75,7 @@ private slots:
     void toggleFirstLastPiecePrioAction();
+
+private:
+    QVariantList getStickyTrackers(const BitTorrent::TorrentHandle* torrent);
```
`QVariantList getStickyTrackers(const BitTorrent::TorrentHandle* torrent) const;`

glassez replied:
Please see my previous comment.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -835,3 +856,91 @@ void TorrentsController::removeCategoriesAction()
+{
+    static const QString working("Working");
+    static const QString disabled("Disabled");
+    static const QString privateMsg("This torrent is private");
```
Remove `static`s, no point doing so IMO.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -835,3 +856,91 @@ void TorrentsController::removeCategoriesAction()
+    QVariantList stickyTrackers;
+
+    QVariantMap dht;
+    dht[KEY_TRACKER_URL] = "** [DHT] **";
```
Not accustomed to list initialization? 
```c++
QVariantMap dht = {
    {KEY_TRACKER_URL, "** [DHT] **"},
    // ...
};
```

Piccirello replied:
I couldn't quite figure out the syntax for this. Thanks!

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -835,3 +856,91 @@ void TorrentsController::removeCategoriesAction()
+    lsd[KEY_TRACKER_DOWNLOADED] = "";
+
+    if (BitTorrent::Session::instance()->isDHTEnabled() && !torrent->isPrivate())
+        dht[KEY_TRACKER_STATUS] = "Working";
```
?? you're not using the constant defined above...

Chocobo1 replied:
also I would write it:
```c++
dht[KEY_TRACKER_STATUS] = (BitTorrent::Session::instance()->isDHTEnabled() && !torrent->isPrivate())
    ? "Working"
    : "Disabled";
```

Piccirello replied:
Definite goof on my end

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -835,3 +856,91 @@ void TorrentsController::removeCategoriesAction()
+    lsd[KEY_TRACKER_SEEDS] = seedsLSD;
+    lsd[KEY_TRACKER_PEERS] = peersLSD;
+
+    stickyTrackers << dht;
```
move `stickyTrackers` declaration above this line and initialize it, seems you're following C rules in modern C++ code... :\

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -835,3 +856,91 @@ void TorrentsController::removeCategoriesAction()
+    }
+
+    uint seedsDHT = 0, seedsPeX = 0, seedsLSD = 0, peersDHT = 0, peersPeX = 0, peersLSD = 0;
+    foreach (const BitTorrent::PeerInfo &peer, torrent->peers()) {
```
use range-based `for` loop.

Chocobo1 commented on src/webui/www/private/scripts/prop-trackers.js:
```diff
@@ -86,11 +86,14 @@ var loadTrackersData = function() {
-                    row[1] = tracker.status;
-                    row[2] = tracker.num_peers;
-                    row[3] = escapeHtml(tracker.msg);
+                    row[0] = tracker.tier;
```
Not sure, how about: 
```javascript
row = [
    tracker.tier,
    // ...
];
```

zeule commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -55,8 +55,12 @@
+const char KEY_TRACKER_TIER[] = "tier";
 const char KEY_TRACKER_MSG[] = "msg";
-const char KEY_TRACKER_PEERS[] = "num_peers";
+const char KEY_TRACKER_RECEIVED[] = "num_peers";
```
TRACKER_RECEIVED == "num_peers" ?

Piccirello replied:
The value associated with this key is used to display the tracker "Received" count, but for legacy reasons the key is `num_peers`. We can't change this without breaking the api.

zeule replied:
Maybe you'd better add four new keys and leave "num_peers" for a deprecation later on?

zeule commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -292,7 +296,11 @@ void TorrentsController::propertiesAction()
 //   - "status": Tracker status
-//   - "num_peers": Tracker peer count
+//   - "tier": Tracker tier
+//   - "num_peers": Tracker received count
```
Received count of what?

Piccirello replied:
Not quite sure, but this is the nomenclature used in the GUI.

Piccirello replied:
I investigated further, it's the total number of peers the torrent is connected to. It's documented [here](https://libtorrent.org/include/libtorrent/torrent_status.hpp) for `torrent_status::numPeers`.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -835,3 +856,91 @@ void TorrentsController::removeCategoriesAction()
         BitTorrent::Session::instance()->removeCategory(category);
 }
+
+QVariantList TorrentsController::getStickyTrackers(const BitTorrent::TorrentHandle* torrent)
```
const BitTorrent::TorrentHandle *torrent

glassez replied:
Does this function depend on TorrentsController? If not, it would be better to have it in an anonymous namespace instead.

Piccirello replied:
Will move it to an anonymous namespace

Piccirello commented:
All comments addressed. These new changes need more testing, will update once complete.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -124,6 +130,80 @@ namespace
+    {
+        const QString working {QLatin1String("Working")};
+        const QString disabled {QLatin1String("Disabled")};
+        const QString privateMsg {QLatin1String("This torrent is private")};
```
Seems these strings need to be translated, try `QCoreApplication::translate("TrackerListWidget", "This torrent is private");` see if it works.

Piccirello replied:
The translation works now.

glassez replied:
Hell, I really don't like the tendency of recent changes to return translated text from an API (and generally, return anything visible directly to the user). You generate unnecessary dependencies again. But I guess you all don't care...

Chocobo1 replied:
@Piccirello 
Or maybe return some constant (integers?) here and do the translation at client side? However I'm not sure how to fetch translated strings at javascript... or this need some html tricks...

glassez replied:
>However I'm not sure how to fetch translated strings at javascript...

JavaScript sources are translatable as well. You can easily create error code to description text mapping there.

glassez replied:
>create error code to description text mapping

Not only **error** codes. Any kind of **status** codes.
>Or maybe return some constant (integers?) 

Integers aren't required, of course. They can be some string codes (e.g. "working", "paused" etc.).

Piccirello replied:
I’m having trouble following, what is the actual recommendation here? Unless someone feels very strongly, I’m inclined to leave these as is.

glassez replied:
The point is that the API is a data source and it should provide data in the form of their values, not their specific representation. Client app handles received data and provide it to the user in form of particular representation (e.g. API provide tracker status as [working; disabled] and then client app show it as translated string, or some color, or icon etc.).

Piccirello replied:
Wait, all of the APIs in `torrentscontroller` already return translated strings. This is a non-issue.

I will change the API back to return translated strings.

glassez replied:
>Wait, all of the APIs in torrentscontroller already return translated strings. This is a non-issue.

Damn! This is an old issue.
>I will change the API back to return translated strings.

Why not just fix old ones? We need to learn from other people's mistakes, not repeat them.

Well, if everyone else thinks my arguments are worthless, you can go on as you like. @Chocobo1, @zeule, @thalieht?

Piccirello replied:
You really like jumping on my PRs to report your issues with the general qbittorent architecture. Please, open a separate PR fixing the issues that you see. It should not hold back new features that follow the existing architecture.

Chocobo1 replied:
>Why not just fix old ones? We need to learn from other people's mistakes, not repeat them.

But changing it now will surely break API... We should enforce correct design for new code, but the issue here is old, personally I would turn a blind eye for this specific case.

@Piccirello 
Just asking for your opinion here, what do you think about the idea (do i18n strings replacement in HTML file) in https://github.com/qbittorrent/qBittorrent/pull/9375#issuecomment-421228476, is it much hassle to do it?

Piccirello replied:
It's easy to do for static strings in HTML files, but more difficult to do for api responses. We end up with code that's very tightly coupled to specific api versions, such as

```
var status;
switch (tracker.status) {
    case "Working":
        status = "QBT_TR(Working)QBT_TR[CONTEXT=TrackerListWidget]";
        break;
    case "Disabled":
        status = "QBT_TR(Disabled)QBT_TR[CONTEXT=TrackerListWidget]";
        break;
    case "This torrent is private":
        status = "QBT_TR(This torrent is private)QBT_TR[CONTEXT=TrackerListWidget]";
        break;
    default:
        status = tracker.status;
}
```

I would think it's considered a feature that the current APIs return translated string values.

Chocobo1 replied:
>It's easy to do for static strings in HTML files, but more difficult to do for api responses. 

Thank you, I hope that doesn't mean impossible for api responses.

>I would think it's considered a feature that the current APIs return translated string values.

I think it *might* be a feature if there are 2 seperate values returned, such as:
```
tracker.status = "working"
tracker.statusI18N = "<translated string>"
```
My point is, if we always return translated strings, then it is cumbersome to write code for the API, as one will need to take locale settings into consideration.

glassez replied:
>I would think it's considered a feature that the current APIs return translated string values.

Disagree.
We should not spawn such flaws where we have the ability to get around it without breaking API. And we have to seriously consider the need to get rid of this "feature" with the next more or less big API update.

In this particular case let it be as is.

>You really like jumping on my PRs to report your issues with the general qbittorent architecture. Please, open a separate PR fixing the issues that you see. It should not hold back new features that follow the existing architecture.

It's nothing personal. This is our usual practice.
Unfortunately, our resources are too limited to allow us to make major architectural changes at a time. None of the project members knows exactly all the code (especially if it has a long history). We often detect some problems in some parts of the code affected by some (possible unrelated) changes. And we try to solve them along the way, if possible (usually offer to do it to the author of the affected PR).

Piccirello replied:
Let's create issues for these large architectural changes, rather than holding up individual features. In certain cases it might be reasonable, but there's no point in holding back a new feature because we can't decide if the API should return translated strings. Just one example, but the point being that we can still make incremental progress while acknowledging large changes that need to occur.

WolfganP commented:
@Piccirello thx for these continuous improvements to the WebUI. Is it possible to implement the "remove tracker" functionality in the WebUI?

Piccirello commented:
@WolfganP It shouldn't be too difficult to implement the entire tracker context menu. Possibly in this PR if I can get it done quickly enough.
<img width="271" alt="screen shot 2018-08-31 at 2 06 09 pm 2" src="https://user-images.githubusercontent.com/8296030/44929083-bfeb5180-ad27-11e8-8a1c-5ab9a07648b6.png">

I also plan to target having trackers in the sidebar, but that may take a bit longer.

zero77 commented:
@Piccirello 
I don’t want to cause problems as you clearly have put a lot of time and effort in to the current system. But, whilst you are working on trackers I think it would be nice to see whether the connection to the tracker is encrypted or not like the E flag next to pears.

Piccirello commented:
@zero77 I'm not sure that libtorrent exposes this information directly. The [announce_entry](https://libtorrent.org/reference-Core.html#announce-entry) doesn't contain any info related to encryption, though tracker traffic is [usually unencrypted](https://stackoverflow.com/questions/31080484/are-libtorrent-transfers-by-default-encrypted) anyway. You can try to infer this yourself by seeing if any of the tracker urls being with `https`.

Piccirello commented:
Context menu implemented. I've also finished my testing of these new features.

Please review the new commits.

zero77 commented:
@Piccirello
Thanks anyway.

Piccirello commented:
Added one commit to implement the changes from #9461

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,15 +626,83 @@ void TorrentsController::addTrackersAction()
+    torrent->addTrackers(trackers);
+}
+
+void TorrentsController::editTrackerAction()
```
Why do you need "edit tracker" abstraction? Since there are no more properties of the tracker that determine its individuality among other trackers of the same torrent, it is enough to have only "add" and "remove" actions.
Let's compare set of integers and set of objects (having several properties). In tge first case we can just remove one int and add another (like in your case), but in the second case we should have separate "edit" action since "remove" + "add" can have different side effect.

zeule replied:
Probably to make the change atomic.

glassez replied:
>Probably to make the change atomic.

Well, it can be acceptable reason in case of Web API. But the correct name for this action is "replace".
The more atomic action is replace  current set of trackers with the new one.

Piccirello replied:
> But the correct name for this action is "replace".

This is an implementation detail that shouldn't be exposed to the user. To them, they're "editing" a specific tracker, which also aligns with the nomenclature in the GUI.

Piccirello replied:
The API is exposed to users as well...

Piccirello commented:
> > If no more comments, please approve
> 
> There is at least one old unresolved conversation (about translated strings). It won't go away on its own just because you're ignoring it.

Quoting my reply:

> I’m having trouble following, what is the actual recommendation here? Unless someone feels very strongly, I’m inclined to leave these as is.


And regarding your recommendation:

> JavaScript sources are translatable as well. You can easily create error code to description text mapping there.

How can this be accomplished? I don't know how to do this. Please do not suggest returning an integer because that is an unacceptable compromise. For examples, see other qBittorrent web APIs and many other external APIs from large companies.

Chocobo1 commented:
>How can this be accomplished? I don't know how to do this. 

Using the same expression as used in HTML files, example: `QBT_TR(URL)QBT_TR[CONTEXT=TrackerListWidget]`.
However this won't work correctly since currently all JS files are cached in client and I don't think JS files should contain mutable data.

>Please do not suggest returning an integer because that is an unacceptable compromise. 

All in all `const QString working {QCoreApplication::translate("TrackerListWidget", "Working")};` should revert back to `const QString working {"Working"};`
And (ideally) the i18n translation replacement should occur in embedded JS in HTML file.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,15 +626,83 @@ void TorrentsController::addTrackersAction()
+    for (QString urlStr : params()["urls"].split('\n')) {
+        urlStr = urlStr.trimmed();
+        QUrl url(urlStr);
+        if (url.isValid())
```
`if (QUrl(urlStr).isValid())`

Piccirello commented on src/gui/properties/trackerlistwidget.h:
```diff
@@ -54,7 +54,7 @@ class TrackerListWidget : public QTreeWidget
         COL_URL,
         COL_STATUS,
-        COL_RECEIVED,
+        COL_CONNECTIONS,
```
This was changed because the column name "Received" is confusing. No one knows what it means.

libtorrent refers to this data as "peers," but this is different than what a user normally thinks of as a "peer."

Chocobo1 replied:
IMO `connections` could be wrongly read as on-going connection count, which is not true.

Chocobo1 replied:
Hmm, utorrent doesn't have this field...

Chocobo1 replied:
>Shall we give it some time and see if anyone objects?

lets proceed!

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,15 +626,83 @@ void TorrentsController::addTrackersAction()
+        throw APIError(APIErrorType::NotFound);
+
+    QUrl origTrackerUrl(origUrl);
+    QUrl newTrackerUrl(newUrl);
```
add `const` to these vars.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,15 +626,83 @@ void TorrentsController::addTrackersAction()
+    if (!newTrackerUrl.isValid())
+        throw APIError(APIErrorType::BadParams, "New tracker URL is invalid");
+
+    QList<BitTorrent::TrackerEntry> trackers = torrent->trackers();
```
`const` here

Piccirello replied:
This can't be const because we modify `trackers`'s members

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,15 +626,82 @@ void TorrentsController::addTrackersAction()
+        if (trackerUrl == origTrackerUrl) {
+            BitTorrent::TrackerEntry newTracker(newTrackerUrl.toString());
+            newTracker.setTier(tracker.tier());
+            match = true;
```
nitpick, I would move this line above `newTracker` declaration.

thalieht commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,15 +626,82 @@ void TorrentsController::addTrackersAction()
+    for (QString urlStr : params()["urls"].split('\n')) {
+        urlStr = urlStr.trimmed();
+        if (QUrl(urlStr).isValid())
+            trackers << urlStr;
```
The [QUrl docs](http://doc.qt.io/qt-5/qurl.html#QUrl-2) say that it fixes any mistakes on the string but you're passing the original string. Isn't it possible that the QUrl constructor will fix some problems that would make isValid() false?

Piccirello replied:
Good point, I'll use the QUrl string.

glassez commented:
>Please do not suggest returning an integer because that is an unacceptable compromise.

My previous comnent:
>Integers aren't required, of course. They can be some string codes (e.g. "working", "paused" etc.).

>How can this be accomplished? I don't know how to do this.

Pseudo code:
```
function tr(source)
{
    switch(source) {
    case "working": return QBT_TR(Working)QBT_TR[CONTEXT=TrackerListWidget];
    // and so on
    }
}
```

glassez commented:
>For examples, see other qBittorrent web APIs and many other external APIs from large companies.

I have seen very bad web interfaces from very large and cool companies (any sophomore from some IT Department could do better without straining too much).

I've repeated again my basic idea:
>The point is that the API is a data source and it should provide data in the form of their values, not their specific representation. Client app handles received data and provide it to the user in form of particular representation (e.g. API provide tracker status as [working; disabled] and then client app show it as translated string, or some color, or icon etc.).

How can some Web client represent tracker status by coloring its name, for example, if you don't return the status in processable form but just its particular representation?

It seems that you are constantly developing in reverse order, adjusting the API to your specific interface. This is the wrong workflow!

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,15 +626,82 @@ void TorrentsController::addTrackersAction()
+            throw APIError(APIErrorType::Conflict, "New tracker URL already exists");
+        if (trackerUrl == origTrackerUrl) {
+            BitTorrent::TrackerEntry newTracker(newTrackerUrl.toString());
+            newTracker.setTier(tracker.tier());
```
I've noticed that you preserve tracker "tier" so let it be "edit" action.

Piccirello commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,15 +626,82 @@ void TorrentsController::addTrackersAction()
+    checkParams({"hash", "urls"});
+
+    const QString hash = params()["hash"];
+    const QStringList urls = params()["urls"].split('|');
```
`addTrackersAction` uses new line as a delimiter, which is pretty non-standard. Almost all our APIs use a pipe, including `removeTrackersAction`.

glassez replied:
Then wrong things should be fixed, isn't it?

Piccirello replied:
That’s a breaking API change not part of this PR.

Piccirello replied:
Feel free to open a PR fixing it.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -630,10 +630,10 @@ void TorrentsController::addTrackersAction()
-        if (QUrl(urlStr).isValid())
-            trackers << urlStr;
+    for (const QString &urlStr : params()["urls"].split('\n')) {
+        auto url = QUrl(urlStr.trimmed());
```
`const QUrl url {urlStr.trimmed()};`

Piccirello commented:
Squashed all fixups

thalieht commented on src/gui/properties/trackerlistwidget.cpp:
```diff
@@ -591,7 +591,7 @@ QStringList TrackerListWidget::headerLabels()
         , tr("URL")
         , tr("Status")
-        , tr("Received")
+        , tr("Tracker Peers")
```
I wonder if it's just me that think this is even more confusing than "Received". I mean because there is already a Peers column and people will wonder "huh? why there are 2 peer columns and which is which. Which is the one i want?"

Piccirello replied:
Maybe `Tracker Connections`? But then this might mislead users into thinking they're connected to all of those peers. Then again, `Seeds` and `Peers` should make it clear how many peers they're connected to.

I think all of this confusion stems from the ambiguous use of `peers` to generally mean other bittorrent clients, and also to mean other users downloading the file (w/ <100% progress).

glassez replied:
>users downloading the file (w/ <100% progress).

Isn't "leech" mean that?
"Peer" is any client, AFAIK.

Piccirello replied:
> Isn't "leech" mean that?

YES, I couldn't think of this word but I think it solves this issue.

I propose the columns should be `peers`, `seeds`, and `leeches`. It should resolve the confusion and is the most "correct."

thalieht replied:
But _peers_ imply all the peers the tracker knows of, doesn't it? Personally i can't think of something short and more descriptive than _Received_ but i'm biased because i already know what it refers to...

Piccirello replied:
libtorrent refers to this field as `num_peers`, describing it as "the number of peers this torrent currently is connected to" [[source](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/torrent_status.hpp#L328)].

thalieht replied:
// tells how many peers the tracker returned in this response. This is
		// not expected to be greater than the ``num_want`` settings. These are not necessarily
		// all new peers, some of them may already be connected.
		int num_peers;

Piccirello replied:
Can anyone else chime in on whether they find this confusing? I would like to go ahead with `peers`, `seeds`, and `leeches` but we can revisit if others disagree.

WolfganP replied:
I'm OK with peer/seeds/leeches. Thx for this functionality!

thalieht commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -320,10 +406,22 @@ void TorrentsController::trackersAction()
+#if LIBTORRENT_VERSION_NUM >= 10000
+        trackerDict[KEY_TRACKER_SEEDS_COUNT] = (tracker.nativeEntry().scrape_complete > 0) ? tracker.nativeEntry().scrape_complete : 0;
+        trackerDict[KEY_TRACKER_PEERS_COUNT] = (tracker.nativeEntry().scrape_incomplete > 0) ? tracker.nativeEntry().scrape_incomplete : 0;
+        trackerDict[KEY_TRACKER_DOWNLOADED_COUNT] = (tracker.nativeEntry().scrape_downloaded > 0) ? tracker.nativeEntry().scrape_downloaded : 0;
```
We changed these in e2b46f01bb19b7faba0f4b3dd050115562ac16c3

Piccirello replied:
I've implemented this change client side (in `prop-trackers.js`). It seems more reasonable to me that this API would return numbers rather than strings, but am open to discuss if you feel strongly.

Piccirello replied:
I suppose this still doesn't disambiguate an unconnected tracker from a connected tracker with 0 seeds/peers. I'll move the change to the API.

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -110,20 +115,94 @@ namespace
     using Utils::String::parseTriStateBool;
 
-    void applyToTorrents(const QStringList &hashes, const std::function<void (BitTorrent::TorrentHandle *torrent)> &func)
+    void applyToTorrents(const QStringList &hashes, const std::function<void (BitTorrent::TorrentHandle *const torrent)> &func)
```
Please revert this change (and all the related changes). It's not required by your feature and it's questionable (absolutely meaningless for me personally) to apply it in passing.

Piccirello replied:
I don't see any reason we wouldn't want this change. In the future I'll move unrelated changes to their own PR; this commit is already a part of this so no point in moving now.

glassez replied:
>I don't see any reason we wouldn't want this change. 

The reason is at least the changes in this line makes functor declaration to be confusing (having "const" applied to parameter passed by value in function declaration is meaningless and confusing).
Having "const" in the following places of this function has minimal meaning since the function code is too trivial and it isn't supposed to become any more complex in the future (but there are no errors formally and some adherents of the "paranoid" safety can insist on its use). The same can be said about the definition of methods that we pass in "applyToTorrents()" when calling (usually it is some one-liner).
Well, I'm not going to insist on the revert of the formally correct changes, if others deem them acceptable. But I insist on never having "const" applied to "byval" parameters in function declaration.

Piccirello replied:
> But I insist on never having "const" applied to "byval" parameters in function declaration.

On this point we agree

> but there are no errors formally and some adherents of the "paranoid" safety can insist on its use

That's mainly my goal here. The point *can* be const, so why not enforce it.

glassez replied:
Ok, fix this one and let's proceed.

Piccirello replied:
Sorry, I don't follow. Is your issue with this line `BitTorrent::TorrentHandle *const torrent`?

glassez replied:
Yes.
"std::function" template argument is similar to function declaration so it should be affected by the restriction above.

Piccirello replied:
But we aren't passing the torrent by value, we're passing a pointer to it.

glassez replied:
>we're passing a pointer to it

Yes. We're passing "pointer to torrent" by value.

Piccirello replied:
Ah I see. What would the syntax be to have the pointer passed by pointer/reference?

glassez replied:
Pointer by reference: `TorrentHandle *&torrent`.
"Pointer by pointer" is incorrect term. You can just have "pointer to pointer" (or even "pointer to pointer to pointer ... to pointer") and pass it either by value or by reference.

glassez commented:
Can tracker manipulations be applied to set of torrents (not to individual one only)?
Many users have been asking for this feature recently, but there is no one to implement it correctly in GUI. There shouldn't be a big problem to do this here (at least in the API). Maybe not in this PR...

Piccirello commented:
Would the goal be to allow adding a list of trackers to multiple torrents at once?

WolfganP commented:
As always, thx @Piccirello for improving the WebUI. BTW, are the columns "sortable"? (ie be able to click on the column header and make the list to sort by that column criteria)

Piccirello commented:
@WolfganP the columns are not currently sortable or resizable. I don’t think it would be that much effort to add that functionality. I will look into it.

glassez commented:
>Would the goal be to allow adding a list of trackers to multiple torrents at once?

I think Yes.
>I'm not sure what other methods could accept multiple torrents.

As far as I remember, users asked for the ability to remove the tracker from multiple torrents.
And then why not allow to replace tracker (aka "edit" here) for multiple torrents?
The intention is the user selects multiple torrents in some UI and it manipulates a common set of their trackers.

Piccirello commented:
@glassez I've created #9780 to track this feature

Piccirello commented:
I haven't found the time to make the Trackers table sortable/resizable. I'd rather get this PR into 4.1.4 and target a sortable/resizable Trackers table and Content table in a future PR.

@glassez @Chocobo1 please review

Chocobo1 commented on src/webui/www/private/scripts/prop-trackers.js:
```diff
@@ -82,21 +85,25 @@ var loadTrackersData = function() {
+                        tracker.num_peers,
+                        (tracker.num_seeds >= 0) ? tracker.num_seeds : "N/A",
+                        (tracker.num_leeches >= 0) ? tracker.num_leeches : "N/A",
+                        (tracker.num_downloaded >= 0) ? tracker.num_downloaded : "N/A",
```
Now the `"N/A"` strings can be translated by wrapping in QBT_TR

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -113,17 +118,91 @@ namespace
         }
     }
+
+    QVariantList getStickyTrackers(const BitTorrent::TorrentHandle *const torrent)
```
It seems we can make it a few lines shorter without sacrificing too much readability:
```c++
    QVariantList getStickyTrackers(const BitTorrent::TorrentHandle *const torrent)
    {
        uint seedsDHT = 0, seedsPeX = 0, seedsLSD = 0, leechesDHT = 0, leechesPeX = 0, leechesLSD = 0;
        for (const BitTorrent::PeerInfo &peer : torrent->peers()) {
            if (peer.isConnecting()) continue;

            if (peer.fromDHT()) {
                if (peer.isSeed())
                    ++seedsDHT;
                else
                    ++leechesDHT;
            }
            if (peer.fromPeX()) {
                if (peer.isSeed())
                    ++seedsPeX;
                else
                    ++leechesPeX;
            }
            if (peer.fromLSD()) {
                if (peer.isSeed())
                    ++seedsLSD;
                else
                    ++leechesLSD;
            }
        }

        const QString working {QCoreApplication::translate("TrackerListWidget", "Working")};
        const QString disabled {QCoreApplication::translate("TrackerListWidget", "Disabled")};
        const QString privateMsg {QCoreApplication::translate("TrackerListWidget", "This torrent is private")};

        const QVariantMap dht {
            {KEY_TRACKER_URL, "** [DHT] **"},
            {KEY_TRACKER_TIER, ""},
            {KEY_TRACKER_MSG, (torrent->isPrivate() ? privateMsg : "")},
            {KEY_TRACKER_STATUS, ((BitTorrent::Session::instance()->isDHTEnabled() && !torrent->isPrivate()) ? working : disabled)},
            {KEY_TRACKER_PEERS_COUNT, 0},
            {KEY_TRACKER_DOWNLOADED_COUNT, 0},
            {KEY_TRACKER_SEEDS_COUNT, seedsDHT},
            {KEY_TRACKER_LEECHES_COUNT, leechesDHT}
        };

        const QVariantMap pex {
            {KEY_TRACKER_URL, "** [PeX] **"},
            {KEY_TRACKER_TIER, ""},
            {KEY_TRACKER_MSG, (torrent->isPrivate() ? privateMsg : "")},
            {KEY_TRACKER_STATUS, ((BitTorrent::Session::instance()->isPeXEnabled() && !torrent->isPrivate()) ? working : disabled)},
            {KEY_TRACKER_PEERS_COUNT, 0},
            {KEY_TRACKER_DOWNLOADED_COUNT, 0},
            {KEY_TRACKER_SEEDS_COUNT, seedsPeX},
            {KEY_TRACKER_LEECHES_COUNT, leechesPeX}
        };

        const QVariantMap lsd {
            {KEY_TRACKER_URL, "** [LSD] **"},
            {KEY_TRACKER_TIER, ""},
            {KEY_TRACKER_MSG, (torrent->isPrivate() ? privateMsg : "")},
            {KEY_TRACKER_STATUS, ((BitTorrent::Session::instance()->isLSDEnabled() && !torrent->isPrivate()) ? working : disabled)},
            {KEY_TRACKER_PEERS_COUNT, 0},
            {KEY_TRACKER_DOWNLOADED_COUNT, 0},
            {KEY_TRACKER_SEEDS_COUNT, seedsLSD},
            {KEY_TRACKER_LEECHES_COUNT, leechesLSD}
        };

        return QVariantList {dht, pex, lsd};
    }
```

Piccirello replied:
This version looks much cleaner. Will change

Piccirello commented:
Really hope we can get this in before 4.1.5 is cut

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -113,17 +118,84 @@ namespace
+            }
+        }
+
+        const QString working {QCoreApplication::translate("TrackerListWidget", "Working")};
```
Well you will not convince me that this is the correct way! Haha, it depends on the translations from the GUI components!.. However, I don't want to open your eyes anymore so you can ignore this comment.

Piccirello replied:
Addressed this in my latest commit. Translation of `status` is now done client-side. The `msg` is still translated server-side since in most cases it's not actually translated at all. It might even make sense to remove translation of the `msg` field entirely.

glassez replied:
>It might even make sense to remove translation of the msg field entirely.

Real trackers can return untranslated messages so it doesn't make much sense to translate this message.
Otherwise, if you still want to do it, I wouldn't rely on translation from GUI and have own one.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -320,11 +395,23 @@ void TorrentsController::trackersAction()
+            {KEY_TRACKER_STATUS, status},
+            {KEY_TRACKER_PEERS_COUNT, data.numPeers},
+            {KEY_TRACKER_MSG, data.lastMessage.trimmed()},
+#if LIBTORRENT_VERSION_NUM >= 10000
```
Do we support libtorrent < 1.0?

glassez replied:
>Do you know?

We don't support libtorrent < 1.0.
I just thought you had the wrong version number here...

Piccirello replied:
This was copied over from `trackerlistwidget.cpp`

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -527,31 +614,98 @@ void TorrentsController::addTrackersAction()
+    }
+
+    if (remainingTrackers.size() == trackers.size())
+        throw APIError(APIErrorType::NotFound);
```
I would throw Conflict error instead. The resource affected by this request is torrent. So it's correct to throw NotFound error only in case of missing torrent.

Piccirello commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -113,17 +118,84 @@ namespace
+        for (const BitTorrent::PeerInfo &peer : torrent->peers()) {
+            if (peer.isConnecting()) continue;
+
+            if (peer.fromDHT()) {
```
Changed to check `peer.isSeed()` before checking dht, pex, and lsd

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -136,28 +136,27 @@ namespace
 
-            if (peer.fromDHT()) {
-                if (peer.isSeed())
+            if (peer.isSeed()) {
```
I wonder if the following is possible:
```c++
if (peer.fromDHT())
    ++(peer.isSeed()? seedsDHT : leechesDHT);
```
Can you test it?

Piccirello replied:
@glassez I don't think it's worth sacrificing readability to save a couple lines

Piccirello replied:
@Chocobo1 Are we certain that a peer cannot come from multiple sources here? Maybe that doesn't make sense (please tell me), but the `peer_info` docs[0] state "the flags indicating which sources a peer can have come from. A peer may have been seen from multiple sources". It's very possible that doesn't include dht/lsd/pex, but just want to be certain.

[0] https://www.libtorrent.org/include/libtorrent/peer_info.hpp

glassez replied:
>I don't think it's worth sacrificing readability to save a couple lines

Ok.

Chocobo1 replied:
>@Chocobo1 Are we certain that a peer cannot come from multiple sources here? 

No, my example code is wrong.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -529,31 +598,98 @@ void TorrentsController::addTrackersAction()
-        torrent->addTrackers(trackers);
     }
+    if (!match)
+        throw APIError(APIErrorType::NotFound);
```
"Conflict"

Chocobo1 commented:
You have conflicts now.

Piccirello commented:
Fixed conflicts

Piccirello commented:
My local builds seem to be failing on the same `asConst` error as some of the travis-ci builds. Anyone know why? @glassez @Chocobo1 @thalieht

thalieht commented:
Looks like g++ 4.8.x has a problem with it https://stackoverflow.com/a/36970437
Should i rename one of the asConst templates until whatever needs gcc 4.8 is not supported? If yes rename to moveAsConst?
Travis was broken at the time so we didn't catch it...

Piccirello commented:
> Looks like g++ 4.8.x has a problem with it

Updating to gcc 4.9 fixed it for me 👍

Chocobo1 commented:
LGTM, wait for other's review before you squash the commits.

Chocobo1 commented:
Please squash commits.

Piccirello commented:
Fixup commits were squashed

Piccirello commented:
Fixed conflict

Chocobo1 commented:
Thank you!
